sympy/polys/numberfields/forms.py

```diff
@@ -0,0 +1,449 @@
+            else:
+                f = self.numerator()
+                # TODO: check whether this is an efficient pow-mod operation or not:
+                g = f ** a % self.T
+                return StandardRep.from_poly(self.T, g) // (self.denom ** a)
+        return NotImplemented
+
+    def __floordiv__(self, a):
+        if is_rat(a):
+            a = sympify(a)
+            if a == 0:
+                raise ZeroDivisionError
+            return self * (1/a)
+        return NotImplemented
+
+    @property
+    def inverse(self):
+        if self._inverse is None:
+            f = self.poly()
+            e, h = f.invert(self.T).clear_denoms()
+            self._inverse = StandardRep.from_poly(self.T, h) // e
+        return self._inverse
+
+    def __rfloordiv__(self, a):
+        return a * self.inverse
+
+    def __mod__(self, a):
+        """
+        Reducing mod an integer a reduces all numerator coeffs mod d*a, where
+        d is our denominator.
+
+        For example, if we represent
+
+            15*x + 1     x + 1
+            --------  =  ----- + 7*x
+               2           2
+
+        then reducing mod 7 should mean throwing away that part that is a poly
+        divisible by 7. But this is achieved by reducing our coeffs mod 2*7.
+        """
+        if is_int(a):
+            a = sympify(a)
+            m = a * self.denom
+            return StandardRep(self.T, [b % m for b in self.coeffs], self.denom)
+        return NotImplemented
+
+    def numerator(self, x=None):
+        """Obtain the numerator as a polynomial over ZZ."""
+        x = x or self.T.gen
+        return Poly(reversed(self.coeffs), x, domain=ZZ)
+
+    def poly(self, x=None):
+        """Obtain the number as a polynomial over QQ."""
+        return self.numerator(x=x) // self.denom
+
+    def norm(self, T=None):
+        """Compute the norm of this number."""
+        T = T or self.T
+        x = T.gen
+        A = self.numerator(x=x)
+        return T.resultant(A) // self.denom**self.n
+
+
+@public
+class HNF(ModuleWithDenominator):
+    """
+    Hermite Normal Form representation for orders and ideals.
+
+    Consists of:
+      - a monic irreducible polynomial T(x) over ZZ
+      - a square, upper triangular n x n matrix W with integer entries,
+        where n = deg(T)
+      - a positive integer d
+    Each column equates to the standard representation of an algebraic number,
+    with the single denominator d being common to all columns. The representations
+    are thus over the power basis in theta, a root of T.
+    """
+
+    def __init__(self, T, W, d):
+        if not W.is_square or not W.is_upper:
+            raise ValueError('HNF matrix must be square and upper triangular')
+        super().__init__(W, d)
+        self.T = T
+        self._mult_tab = None
+
+    @classmethod
+    def for_power_basis(cls, T):
+        from sympy.matrices import Matrix
+        return cls(T, Matrix.eye(T.degree()), 1)
```

Comment:
The polys module should not use the matrices module. This should be DomainMatrix.